notebooks/Solutions/Causal Interpretation for Employee Attrition Dataset.ipynb

```diff
@@ -0,0 +1,954 @@
+    "\n",
+    "categorical_transformer = Pipeline(\n",
+    "    steps=[\n",
+    "        (\"imputer\", SimpleImputer(strategy=\"constant\", fill_value=\"missing\")),\n",
+    "        (\"onehot\", OneHotEncoder(handle_unknown=\"error\", drop=\"first\")),\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "transformations = ColumnTransformer(\n",
+    "    transformers=[\n",
+    "        (\"num\", numeric_transformer, numerical),\n",
+    "        (\"cat\", categorical_transformer, categorical),\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "# Append classifier to preprocessing pipeline.\n",
+    "# Now we have a full prediction pipeline.\n",
+    "clf = Pipeline(\n",
+    "    steps=[(\"preprocessor\", transformations), (\"classifier\", LGBMClassifier())]\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Train a LightGBM classification model, which you want to explain"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.model_selection import GridSearchCV\n",
+    "\n",
+    "param_grid = {\n",
+    "    \"classifier__learning_rate\": [0.1, 0.05, 0.01],\n",
+    "    \"classifier__max_depth\": [3, 5, 10],\n",
+    "}\n",
+    "search = GridSearchCV(clf, param_grid, n_jobs=-1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'classifier__learning_rate': 0.1, 'classifier__max_depth': 3}"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "search.fit(x_train, y_train)\n",
+    "search.best_params_"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Correlation Interpretation\n",
+    "We explain this ML model by understanding the top important features to predict the employee attrition, internally using shap value."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# get the fitted model and transformer\n",
+    "fitted_model = search.best_estimator_[\"classifier\"]\n",
+    "fitted_transformer = search.best_estimator_[\"preprocessor\"]\n",
+    "# get the feature name after featurization\n",
+    "column_names = numerical.tolist()\n",
+    "column_names += (\n",
+    "    search.best_estimator_[\"preprocessor\"]\n",
+    "    .transformers_[1][1]\n",
+    "    .steps[1][1]\n",
+    "    .get_feature_names(categorical)\n",
+    "    .tolist()\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Passing 1176 background samples may lead to slow runtimes. Consider using shap.sample(data, 100) to create a smaller background data set.\n"
```

Comment:
Should we be heeding this warning and lowering the number of samples?

Reply:
I think this is fine, given 1k samples is not large and the running speed is not slow, we could get a more accurate estimate.